src/main/kotlin/org/rust/lang/core/types/infer/Expressions.kt

```diff
@@ -56,8 +56,20 @@ fun inferExpressionType(expr: RsExpr): Ty {
             val boundMethod = expr.reference.advancedResolve()
             val method = boundMethod?.element as? RsFunction ?: return TyUnknown
 
-            val returnType = (method.retType?.typeReference?.type ?: TyUnit)
-                .substitute(boundMethod.typeArguments)
+            val typeRef = method.retType?.typeReference
+            // if calling method belongs to some trait `Tr`
+            // and struct or enum doesn't have explicit implementation of this trait,
+            // for example, this trait is implemented with `derive` attribute
+            // and return type is `Self`
+            // then `typeRef.type` returns `Tr`
+            // although the real type of expression is type of struct or enum
+            // that implements trait `Tr`
+            val returnType = if (typeRef.isCselfRef) {
```

Comment:
Hm, I think this won't work for a more complicated case like

```
trait BiClone {
    fn bi_clone(&self) -> (Self, Self)
}
```

What we really need is to add `Self -> StructType` mapping to the `boundMethod.typeArguments`, so that the original code of just `.substitute(mapping)` would do conversion of `Self` to an actual type.